Codechange: Performance improvement in k-d tree FindNearest()

GabdaZM opened the pull request:
The improvement is small, around 4-8%.
I have tested it on a few different map by calling the `FindNearest()` function on every tile of a map, and counting the number of the `FindNearestRecursive()` calls.
The version of this PR draft contains the old and the new `FindNearest()` functions, and an additional console command 'count_calls' to test the improvement.

LordAro commented:
Usual commit message stuff applies.

In terms of the review, I'd recommend splitting it up properly - replace the function with the first commit, and add the debug stuff in a separate commit. Makes it much easier to review, especially when the debug stuff changes the contents of the function (removal of const due to `call_counter` - maybe `mutable` might be better here?)

GabdaZM committed:
Codechange: Performance improvement in k-d tree FindNearest()

GabdaZM commented:
Commit message changed.
Split the commit into two according to your suggestion. The commit message of the second one is made to fail the check, just in case.
Changed the counter to be `mutable`, thanks for the tip!

GabdaZM commented:
Added TICC-TOCC to measure time as well. It seems like the performance improvement is closer to 8%.

LordAro commented:
Needs a rebase

Correction: Probably just needs the debug code removing

nielsmh commented on src/core/kdtree.hpp:
```diff
@@ -240,8 +241,9 @@ class Kdtree {
 	}
 	/** Search a sub-tree for the element nearest to a given point */
-	node_distance FindNearestRecursive(CoordT xy[2], size_t node_idx, int level) const
+	node_distance FindNearestRecursive(CoordT xy[2], size_t node_idx, int level, DistT limit = (DistT)(MapSizeX() + MapSizeY())) const
```
I don't really like this having a non-constant default value for a parameter.
Could you change it to use `std::numeric_limits<DistT>::max()` instead?

nielsmh reviewed (dismissed):
The test/measurement code needs to be removed.
